### PR TITLE
Improve the slot and the data sections

### LINKED ISSUE #19: Improve slots docs
From https://github.com/surface-ui/surface/issues/302#issuecomment-804905604:

> Update the docs to inform that slots are always lists and `<slot name="<name>" />` is just a short hand for `<slot name="<name>" index="0" />`

### LINKED ISSUE #20: Update doc to inform that if you need some data to be initialized, you need to add a default value, even if it's nil
Related to https://github.com/surface-ui/surface/issues/192

### PR DESCRIPTION
**Pull request by Malian**
Fixes #19
Fixes #20

### CODE REVIEW

File: lib/surface_site_web/live/slots.ex [hunk 1]
```diff
@@ -49,72 +53,71 @@ defmodule SurfaceSiteWeb.Slots do
                 Slots are similar to properties as they are exposed as part of the component's **public API**.
                 The main difference is that while properties are passed as attributes, slots are injected inside
                 the component's body.
+
+                To declare a `slot`, you must use the `slot` macro and provide a name to the slot.
+                In the example below, the slot `default` is declared and it is required.
```
Comment (msaraiva) [inline lib/surface_site_web/live/slots.ex:58]:
```suggestion
                In the example below, the slot `default` is declared as required.
```

File: lib/surface_site_web/live/slots.ex [hunk 2]
```diff
@@ -49,72 +53,71 @@ defmodule SurfaceSiteWeb.Slots do
...
+                  />
                 </footer>
               </div>
 
               <#Markdown>
-                Now, if the user tries to use a `Hero` without defining any content, a proper
-                `missing required slot "default"` error will be raised at compile-time.
+                By defining any children inside `<slot>...</slot>`, that content becomes the default content.
+                Note that we removed the `required` option in the `slot` declaration. If we had not done so,
+                a warning would inform the user that fallback content has no effect, thanks to the compiler!
```
Comment (msaraiva) [inline lib/surface_site_web/live/slots.ex:112]:
```suggestion
                a warning would inform the user that the fallback content would have no effect, thanks to the compiler!
```

File: lib/surface_site_web/live/slots.ex [hunk 3]
```diff
@@ -131,27 +134,42 @@ defmodule SurfaceSiteWeb.Slots do
...
+                    show_line_numbers
+                    selected_lines="7-9, 11-13, 15-18"
+                  />
                 </footer>
               </div>
 
               <#Markdown>
+                As you can see in the example, we are using 3 slots:
+                - The header slot
+                - The default slot that contains everything that is not in another slot
```
Comment (msaraiva) [inline lib/surface_site_web/live/slots.ex:150]:
```suggestion
                - The default slot that contains everything that is not in any other slot
```

File: lib/surface_site_web/live/slots.ex [hunk 4]
```diff
@@ -131,27 +134,42 @@ defmodule SurfaceSiteWeb.Slots do
...
+                    selected_lines="5,8,11,18,23,27"
+                  />
                 </footer>
               </div>
+              <#Markdown>
+                  > **Note**: Pay attention that defining a `<slot />` without a name is the same as defining it as `<slot name="default"/>`.
+              </#Markdown>
+
 
               <#Markdown>
```
Comment (msaraiva) [inline lib/surface_site_web/live/slots.ex:172]:
```suggestion

```

File: lib/surface_site_web/live/slots.ex [hunk 5]
```diff
@@ -330,8 +372,16 @@ defmodule SurfaceSiteWeb.Slots do
                     but instead of just defining the name of the prop (as we did for our `Rating` component),
                     we bound the value of that prop to each value (item) produced by the generator `items`.
 
-                  1. We use `<slot>` to render each column's content passing the current item.
+                  2. We use `<slot>` to render each column's content passing the current item.
               </#Markdown>
+
+              <#Markdown>
```
Comment (msaraiva) [inline lib/surface_site_web/live/slots.ex:378]:
```suggestion

```

File: lib/surface_site_web/live/slots.ex [hunk 6]
```diff
@@ -330,8 +372,16 @@ defmodule SurfaceSiteWeb.Slots do
...
+                > **Note**: It is the right moment to inform you that slots are always lists
+                and `<slot name="<name>" />` as defined at the beginning of the section
+                is just a short hand for `<slot name="<name>" index="0" />`. It means that, under the hood,
+                `<slot name="<name>" />` will be replaced by the first value of the list `@assigns.<name>`.
```
Comment (msaraiva) [inline lib/surface_site_web/live/slots.ex:382]: I think a shorter version would be enough. Something like:

```
> **Note**: Slot contents are always passed as lists. However, 
if you expect receiving a single content block, like at the beginning
of this section, you can use `<slot name="<name>" />` as a 
shorthand for `<slot name="<name>" index="0" />`. 
```

WDYT?

**Review by msaraiva — COMMENTED** [review]
Hey @Malian! Great job on the docs. I left a couple of minor comments for your consideration. Also, since we're both non-native english speakers, it would be nice to get a final review from any of our american collaborators :)

@lnr0626 and @paulstatezny in case you have the time, I'd appreciate a quick look from at least one of you.

@lnr0626 I'm not sure yet how the changes in https://github.com/surface-ui/surface/pull/283 will affect the docs but I think it's probably a good idea to get this one merged before. Right?

File: lib/surface_site_web/live/slots.ex [hunk 7]
```diff
@@ -49,72 +53,71 @@ defmodule SurfaceSiteWeb.Slots do
...
+                <Hero>
+                  My Custom Content
+                </Hero>
+              </#Example>
 
               <#Markdown>
-                By defining any children inside `<slot>...</slot>`, that content becomes the default content.
-              </#Markdown>
+                If the user tries to use a `Hero` without defining any content, a proper
+                `missing required slot "default"` error will be raised at compile-time.
```
Comment (lnr0626) [inline lib/surface_site_web/live/slots.ex:85]:
```suggestion
                If the user tries to use the `Hero` component without defining any content, a
                `missing required slot "default"` error will be raised at compile-time.
```

File: lib/surface_site_web/live/slots.ex [hunk 8]
```diff
@@ -49,72 +53,71 @@ defmodule SurfaceSiteWeb.Slots do
...
+                    show_line_numbers
+                    selected_lines="5, 11-13"
+                  />
                 </footer>
               </div>
 
               <#Markdown>
-                Now, if the user tries to use a `Hero` without defining any content, a proper
-                `missing required slot "default"` error will be raised at compile-time.
+                By defining any children inside `<slot>...</slot>`, that content becomes the default content.
```
Comment (lnr0626) [inline lib/surface_site_web/live/slots.ex:110]:
```suggestion
                If at least one child element is defined inside `<slot>...</slot>`, the inner content is used as the default content for that slot.
```

File: lib/surface_site_web/live/slots.ex (hunk 2 above)
Comment (lnr0626) [reply lib/surface_site_web/live/slots.ex:112]: For this, should we say `a warning would inform you` instead of `inform the user`?

**Review by lnr0626 — COMMENTED** [review]
Yup, I'd agree on meeting this before the changes from that pr.

I had a few small nits, but other than those this looks good.

File: lib/surface_site_web/live/data.ex [hunk 9]
```diff
@@ -51,6 +51,13 @@ defmodule SurfaceSiteWeb.Data do
                 decrement/increment the counter's value.
               </#Markdown>
 
+              <#Markdown>
+                > **Note**: The `data` macro is also available in `Surface.Component`
+                even if that kind of component does not have state.
+                It allows you to define intermediary assigns in the `update` function of the component
+                that computes values based on one or several properties.
```
Comment (paulstatezny) [inline lib/surface_site_web/live/data.ex:58]: By the way, is `update` highlighted anywhere in the docs? I don't think I've seen it.

**Review by msaraiva — APPROVED** [review]

**Pull request merged**